File: Code/marathon_Practice/Problem1-10/Problem2/solution.c
#include <stdio.h>
#include <stdlib.h>

struct Node{
    int data;
    int rep;
    struct Node* next;
};

struct Node** hashTable = NULL;
int tableSize;
int count = 0;

int hashFn(int key){
    int slot = key % tableSize;
    return slot;
}

int search(int key){
    int slot = hashFn(key);
    struct Node* temp = hashTable[slot];
    while(temp != NULL){
        if(temp->data == key){
            return slot;
        }
        temp = temp->next;
    } 
    return -1;
}

int insert(int key){
    int slot = hashFn(key);
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->rep = 1;
    if(search(key) != -1){
        newNode->rep = newNode->rep + 1;
        if (newNode->rep == 2){
            return key;
        }
    }

    newNode->data = key;
    newNode->next = hashTable[slot]; 
    hashTable[slot] = newNode;
    return -1;
}

int searchRep(int key){
    int slot = hashFn(key);
    struct Node* temp = hashTable[slot];
    while(temp != NULL){
        if(temp->rep == 2){
            return key;
        }
        temp = temp->next;
    } 
    return -1;

}


int main(){

    int n;
    scanf("%d", &n);

    tableSize = n;
    hashTable = (struct Node**)calloc(tableSize, sizeof(struct Node*));

    for(int i = 0; i < n; i++){
        int val;
        scanf("%d", &val);
        if(insert(val) != -1){
            printf("%d",val);
            return val;
        }
    }

    printf("%d",-1);
    return -1;
}

File: Code/marathon_Practice/Problem1-10/Problem2/solution.py
hashtable = {}
size = 1

def hash_fn(key: int) -> int:
    return key % size

def insert(key: int) -> int:
    slot = hash_fn(key)
    if slot not in hashtable:
        hashtable[slot] = []
    hashtable[slot].append(key)
    if len(hashtable[slot]) >= 2:
        return key
    return -1

size = int(input(""))
data = input("")
data = list(map(int, data.split()))

def run(data) -> int:
    for i in data:
        if insert(i) != -1:
            return i 
    return -1

print(run(data))
